backend/dataall/modules/dataset_sharing/services/data_sharing_service.py

```diff
@@ -226,3 +246,108 @@ def revoke_share(cls, engine: Engine, share_uri: str):
             share_sm.update_state(session, share, new_share_state)
 
             return revoked_folders_succeed and revoked_s3_buckets_succeed and revoked_tables_succeed
+
+    @staticmethod
+    def acquire_lock(dataset_uri, session, share_uri):
+        """
+        Attempts to acquire a lock on the dataset identified by dataset_id.
+
+        Args:
+            dataset_uri: The ID of the dataset for which the lock is being acquired.
+            session (sqlalchemy.orm.Session): The SQLAlchemy session object used for interacting with the database.
+            share_uri: The ID of the share that is attempting to acquire the lock.
+
+        Returns:
+            bool: True if the lock is successfully acquired, False otherwise.
+        """
+        try:
+            # Execute the query to get the DatasetLock object
+            dataset_lock = (
+                session.query(DatasetLock)
+                .filter(
+                    and_(
+                        DatasetLock.datasetUri == dataset_uri,
+                        ~DatasetLock.isLocked
+                    )
+                )
+                .first()
+            )
+
+            # Check if dataset_lock is not None before attempting to update
+            if dataset_lock:
+                # Update the attributes of the DatasetLock object
+                dataset_lock.isLocked = True
+                dataset_lock.acquiredBy = share_uri
+
+                session.commit()
+                return True
+            else:
+                log.info("DatasetLock not found for the given criteria.")
+                return False
+
+        except Exception as e:
+            session.rollback()
+            log.error("Error occurred while acquiring lock:", e)
+            return False
+
+    @staticmethod
+    def acquire_lock_with_retry(dataset_uri, session, share_uri):
+        max_retries = 10
+        retry_interval = 60
```

Comment:
These 2 numbers are open to discussion. 
@dlpzx @noah-paige What do you think of these numbers?

Reply:
I think these numbers make sense - will do some testing to get a better sense of user experience but think 10 minute max wait seems good

Should we make these global variables at the top of this file, such as 
```
MAX_RETRIES = 10
RETRY_INTERVAL = 60
```

Reply:
Yes I will make them global variables. 

We could also make the max_retries 20 and keep the retry_interval 30. That would still give us 10 min coverage but only a 30 sec delay between retries. However, this could cause heavy throttling.